lib/rules/no-global-regexp-flag-in-query.ts

```diff
@@ -0,0 +1,94 @@
+import { ASTUtils, TSESTree } from '@typescript-eslint/utils';
+
+import { createTestingLibraryRule } from '../create-testing-library-rule';
+import {
+  isMemberExpression,
+  isCallExpression,
+  isProperty,
+  isObjectExpression,
+  getDeepestIdentifierNode,
+  isLiteral,
+} from '../node-utils';
+
+export const RULE_NAME = 'no-global-regexp-flag-in-query';
+export type MessageIds = 'noGlobalRegExpFlagInQuery';
+type Options = [];
+
+export default createTestingLibraryRule<Options, MessageIds>({
+  name: RULE_NAME,
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: 'Disallow the use of the global RegExp flag (/g) in queries',
+      recommendedConfig: {
+        dom: false,
+        angular: false,
+        react: false,
+        vue: false,
+      },
+    },
+    messages: {
+      noGlobalRegExpFlagInQuery:
+        'Avoid using the global RegExp flag (/g) in queries',
+    },
+    fixable: 'code',
+    schema: [],
+  },
+  defaultOptions: [],
+  create(context, _, helpers) {
+    function report(literalNode: TSESTree.Node) {
+      if (
+        isLiteral(literalNode) &&
+        'regex' in literalNode &&
+        literalNode.regex.flags.includes('g')
+      ) {
+        context.report({
+          node: literalNode,
+          messageId: 'noGlobalRegExpFlagInQuery',
+          fix(fixer) {
+            const splitter = literalNode.raw.lastIndexOf('/');
+            const raw = literalNode.raw.substring(0, splitter);
+            const flags = literalNode.raw.substring(splitter + 1);
+            const flagsWithoutGlobal = flags.replace('g', '');
+
+            return fixer.replaceText(
+              literalNode,
+              `${raw}/${flagsWithoutGlobal}`
+            );
+          },
+        });
+        return true;
+      }
+      return false;
+    }
+
+    return {
+      CallExpression(node) {
+        const identifierNode = getDeepestIdentifierNode(node);
+        if (!identifierNode || !helpers.isQuery(identifierNode)) {
+          return;
+        }
+
+        const [firstArg, secondArg] = isCallExpression(identifierNode.parent)
+          ? identifierNode.parent.arguments
+          : isMemberExpression(identifierNode.parent) &&
+            isCallExpression(identifierNode.parent.parent)
+          ? identifierNode.parent.parent.arguments
+          : [];
```

Comment:
This is not really readable, could you transform it into regular if/else-if/else statements?

I know you won't be able to declare the variables as const, but the code will be much cleaner.

Reply:
💯

Reply:
Thanks for calling me out 😅